File: 第12章-架构感知与定位/04-案例/matlab/gnss_imu_ekf.m
thisDir = fileparts(mfilename('fullpath'));
root = fileparts(fileparts(fileparts(thisDir)));
addpath(fullfile(root, 'common', 'matlab'));
p = vehicle_params();
rng(2);
dt = 0.05; T = 20; n = round(T/dt);
x_true = [0;0;0]; v = 12;
x_hat = [0.5;-0.5;0.1]; P = diag([9,9,0.2]);
Q = diag([0.05,0.05,0.002])*dt; R = diag([4,4]);
C = [1 0 0; 0 1 0];
traj_t = zeros(2,n); traj_g = traj_t; traj_e = traj_t;
for k = 1:n
    delta = deg2rad(4)*sin(2*pi*(k-1)*dt/12);
    x_true = bicycle_euler(x_true, v, delta, dt, p.L);
    r_meas = v/p.L*tan(delta) + 0.02*randn;
    psi = x_hat(3);
    A = [1 0 -v*dt*sin(psi); 0 1 v*dt*cos(psi); 0 0 1];
    x_hat = x_hat + [v*dt*cos(psi); v*dt*sin(psi); r_meas*dt];
    P = A*P*A' + Q;
    y = x_true(1:2) + 2*randn(2,1);
    [x_hat, P] = kf_update(x_hat, P, y, C, R);
    traj_t(:,k)=x_true(1:2); traj_g(:,k)=y; traj_e(:,k)=x_hat(1:2);
end
fprintf('GNSS RMSE=%.3f  EKF RMSE=%.3f\n', ...
    sqrt(mean(sum((traj_g-traj_t).^2,1))), sqrt(mean(sum((traj_e-traj_t).^2,1))));
figure; plot(traj_t(1,:),traj_t(2,:),'k-', traj_g(1,:),traj_g(2,:),'.', traj_e(1,:),traj_e(2,:),'k--');
axis equal; grid on; legend('true','GNSS','EKF'); title('GNSS + yaw-rate EKF');

File: 第12章-架构感知与定位/04-案例/python/gnss_imu_ekf.py
from __future__ import annotations

import sys
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np

ROOT = Path(__file__).resolve().parents[3]
sys.path.insert(0, str(ROOT / "common" / "python"))
from bicycle_kinematic import bicycle_euler, wrap_to_pi  # noqa: E402
from kalman import kf_update  # noqa: E402
from vehicle_params import VEH  # noqa: E402


def main() -> None:
    rng = np.random.default_rng(2)
    dt = 0.05
    T = 20.0
    n = int(T / dt)
    x_true = np.array([0.0, 0.0, 0.0])
    v = 12.0
    x_hat = np.array([0.5, -0.5, 0.1])
    P = np.diag([9.0, 9.0, 0.2])
    Q = np.diag([0.05, 0.05, 0.002]) * dt
    R = np.diag([4.0, 4.0])
    C = np.array([[1.0, 0.0, 0.0], [0.0, 1.0, 0.0]])
    traj_t = np.zeros((n, 2))
    traj_g = np.zeros((n, 2))
    traj_e = np.zeros((n, 2))
    openloop = x_hat.copy()
    traj_o = np.zeros((n, 2))
    for k in range(n):
        delta = np.deg2rad(4.0) * np.sin(2 * np.pi * k * dt / 12.0)
        x_true = bicycle_euler(x_true, v, delta, dt, VEH.L)
        r_meas = (v / VEH.L) * np.tan(delta) + rng.normal(0.0, 0.02)
        # IMU-driven prediction (v, r)
        psi = x_hat[2]
        A = np.array(
            [
                [1.0, 0.0, -v * dt * np.sin(psi)],
                [0.0, 1.0, v * dt * np.cos(psi)],
                [0.0, 0.0, 1.0],
            ]
        )
        x_hat = x_hat + np.array([v * dt * np.cos(psi), v * dt * np.sin(psi), r_meas * dt])
        x_hat[2] = wrap_to_pi(x_hat[2])
        P = A @ P @ A.T + Q
        openloop = openloop + np.array(
            [v * dt * np.cos(openloop[2]), v * dt * np.sin(openloop[2]), r_meas * dt]
        )
        y = x_true[:2] + rng.normal(0.0, 2.0, size=2)
        x_hat, P, _ = kf_update(x_hat, P, y, C, R)
        x_hat[2] = wrap_to_pi(x_hat[2])
        traj_t[k] = x_true[:2]
        traj_g[k] = y
        traj_e[k] = x_hat[:2]
        traj_o[k] = openloop[:2]
    rmse_g = np.sqrt(np.mean(np.sum((traj_g - traj_t) ** 2, axis=1)))
    rmse_e = np.sqrt(np.mean(np.sum((traj_e - traj_t) ** 2, axis=1)))
    print(f"GNSS RMSE={rmse_g:.3f} m  EKF RMSE={rmse_e:.3f} m")

    fig, ax = plt.subplots(figsize=(6, 6))
    ax.plot(traj_t[:, 0], traj_t[:, 1], "k-", label="true")
    ax.plot(traj_g[:, 0], traj_g[:, 1], ".", color="0.6", ms=3, label="GNSS")
    ax.plot(traj_e[:, 0], traj_e[:, 1], "k--", label="EKF")
    ax.set_aspect("equal")
    ax.grid(True)
    ax.legend()
    ax.set_title("GNSS + yaw-rate EKF")
    fig.tight_layout()
    out = Path(__file__).with_name("gnss_imu_ekf.png")
    fig.savefig(out, dpi=120)
    print(f"saved {out}")
    plt.show()


if __name__ == "__main__":
    main()
